### PR TITLE
SISRP-22628 - 6.0 - Backend - Link API for Campus Solutions

### PR DESCRIPTION
**Pull request by dfkaye**
https://jira.berkeley.edu/browse/SISRP-22628

@redconfetti @paulkerschen @raydavis  Link API backend is ready for first-pass review

### CODE REVIEW

File: app/models/campus_solutions/link.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+      normalize_response(response).parsed_response
+    end
+
+    def normalize_response(response)
+      # Make sure response is well-structured.
+      container = response.parsed_response['UC_LINK_RESOURCES']
+      return response if container.blank? || container['IS_FAULT'].present?
+
+      # Insert LINKS element array.
+      container['Links'] = Array.wrap container['Link']
```
Comment (johncrossman) [inline app/models/campus_solutions/link.rb]: @dfkaye the `container` var is not used after this point. Did you intend to put `Links` to the response?

File: app/models/campus_solutions/link.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    def get_url(urlId, options = {})
+      # get cached response
+      response = get
+
+      links = response[:feed][:ucLinkResources][:links]
+      link = links.find do |link|
+        link[:urlId] == urlId
+      end
+
+      if !link.nil? && !options.empty?
```
Comment (johncrossman) [inline app/models/campus_solutions/link.rb]: @dfkaye Perhaps more readable: `if link && options.any?`

File: app/models/campus_solutions/link.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        options.each do |k, v|
+          link[:url] = link[:url].gsub("{#{k}}", v)
+        end
+      end
+
+      response[:feed] = { :link => link }
+
```
Comment (johncrossman) [inline app/models/campus_solutions/link.rb]: @dfkaye I suggest cutting down on number of blank lines. No biggie.

File: config/routes.rb [hunk 4]
```diff
@@ -195,6 +195,7 @@
   get '/api/campus_solutions/fpp_enrollment' => 'campus_solutions/fpp_enrollment#get', :via => :get, :defaults => { :format => 'json' }
   get '/api/campus_solutions/emergency_contacts' => 'campus_solutions/emergency_contacts#get', :via => :get, :defaults => { :format => 'json' }
   get '/api/campus_solutions/residency_message' => 'campus_solutions/residency_message#get', :defaults => { :format => 'json' }
+  get '/api/campus_solutions/link' => 'campus_solutions/link#get', :via => :get, :defaults => { :format => 'json' }
```
Comment (johncrossman) [inline config/routes.rb]: @dfkaye `:via => :get, ` is not necessary

File: spec/controllers/campus_solutions/link_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
+describe CampusSolutions::LinkController do
+  context 'link feed' do
+    let(:feed) { :get }
+
+    it_behaves_like 'an unauthenticated user'
+
+    context 'authenticated user' do
```
Comment (johncrossman) [inline spec/controllers/campus_solutions/link_controller_spec.rb]: @dfkaye do we require that a user be an 'authenticated user' to get links from link API?

File: app/models/campus_solutions/link.rb (hunk 2 above)
Comment (dfkaye) [inline app/models/campus_solutions/link.rb]: +1 @johncrossman

File: app/models/campus_solutions/link.rb (hunk 3 above)
Comment (dfkaye) [inline app/models/campus_solutions/link.rb]: @johncrossman I'll remove them

File: config/routes.rb (hunk 4 above)
Comment (dfkaye) [inline config/routes.rb]: @johncrossman zapped

File: spec/controllers/campus_solutions/link_controller_spec.rb (hunk 5 above)
Comment (dfkaye) [inline spec/controllers/campus_solutions/link_controller_spec.rb]: @johncrossman I merely assumed it - will that cause us problems?

File: app/models/campus_solutions/link.rb (hunk 1 above)
Comment (dfkaye) [inline app/models/campus_solutions/link.rb]: @johncrossman modified +1

**Comment by raydavis** [comment]
@dfkaye , commit log lines like "LINKAPI" make life unnecessarily difficult in a shared branch. Can you please make your commit messages in the following format?

JIRA-CODE A brief description of the commit

e.g.,

```
git commit --amend -m "SISRP-22628 Add CS Link API support"
```

File: app/controllers/campus_solutions/link_controller.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+module CampusSolutions
+  class LinkController < CampusSolutionsController
```
Comment (raydavis) [inline app/controllers/campus_solutions/link_controller.rb]: Why does a "backend" PR expose a passthrough to the CS API?

File: config/routes.rb (hunk 4 above)
Comment (raydavis) [inline config/routes.rb]: Same question as with the controller. Are we sure the front-end needs to hit this endpoint directly?

File: app/controllers/campus_solutions/link_controller.rb (hunk 6 above)
Comment (raydavis) [inline app/controllers/campus_solutions/link_controller.rb]: @dfkaye talked to me about the intent of the JIRA and this makes sense to me now.

File: spec/controllers/campus_solutions/link_controller_spec.rb (hunk 5 above)
Comment (raydavis) [inline spec/controllers/campus_solutions/link_controller_spec.rb]: This is where knowing how the CalCentral endpoint will be used might be handy. FWIW, I'd say that until we see there's a need for a public (not-logged-in-yet) link in the UX, we might as well guard the request.

**Comment by raydavis** [comment]
Thank you for the amendments & the explanation, @dfkaye -- just waiting on old man Travis...

**Pull request merged**